File: frontend/lib/categories.ts
import { api } from "./api";

export type Category = {
  id: number;
  name: string;
};

export async function listCategories(): Promise<Category[]> {
  const res = await api<{ categories: Category[] }>("/categories");
  return res.categories;
}

export async function createCategory(name: string): Promise<Category> {
  return api<Category>("/categories", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
}

export async function updateCategory(id: number, name: string): Promise<Category> {
  return api<Category>(`/categories/${id}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
  });
}

export async function deleteCategory(id: number): Promise<void> {
  await api<void>(`/categories/${id}`, { method: "DELETE" });
}

File: backend/app/routers/categories.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.category import CategoryListResponse, CategoryResponse, CreateCategoryRequest, UpdateCategoryRequest
from app.services.category_service import CategoryService

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get("", response_model=CategoryListResponse)
async def list_categories(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = CategoryService(db)
    return await svc.list_categories(current_user)


@router.post("", response_model=CategoryResponse, status_code=201)
async def create_category(
    body: CreateCategoryRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = CategoryService(db)
    return await svc.create_category(current_user, body)


@router.patch("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: int,
    body: UpdateCategoryRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = CategoryService(db)
    return await svc.update_category(current_user, category_id, body)


@router.delete("/{category_id}", status_code=204)
async def delete_category(
    category_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = CategoryService(db)
    await svc.delete_category(current_user, category_id)
